src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -106,5 +116,95 @@ private IArgumentConstraint CreateParamArrayConstraint(NewArrayExpression expres
 
             return new AggregateArgumentConstraint(result);
         }
+
+        private class ArgumentConstraintExpressionVisitor : ExpressionVisitor
+        {
+            private readonly List<Expression> constraintExpressions;
+            private readonly HashSet<Expression> visitedExpressions;

```

Comment:
> you're forcing a descent down the tree in some cases.

Yes. But now that I think about it, there might be an easier way. Let me try something.